modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java

```diff
@@ -602,4 +612,68 @@ public static boolean typeFamiliesAreCompatible(RelDataTypeFactory typeFactory,
             return false;
         }
     }
+
+    /** Creates an instance of {@link RowSchema} from a list of the given {@link RelDataType}s. */
+    public static RowSchema rowSchemaFromRelTypes(List<RelDataType> types) {
+        RowSchema.Builder fieldTypes = RowSchema.builder();
+
+        for (RelDataType relType : types) {
+            TypeSpec typeSpec = convertToTypeSpec(relType);
+            fieldTypes.addField(typeSpec);
+        }
+
+        return fieldTypes.build();
+    }
+
+    private static TypeSpec convertToTypeSpec(RelDataType type) {
+        boolean simpleType = type instanceof BasicSqlType;
+        boolean nullable = type.isNullable();
+
+        if (type instanceof IgniteCustomType) {
+            NativeType nativeType = IgniteTypeFactory.relDataTypeToNative(type);
+            return RowSchemaTypes.nullableNativeType(nativeType, nullable);
+        } else if (SqlTypeName.ANY == type.getSqlTypeName()) {
+            // TODO: After https://issues.apache.org/jira/browse/IGNITE-19096 is fixed
+            //  it should be possible to remove branch.
+            // TODO Some JSON functions that return ANY as well : https://issues.apache.org/jira/browse/IGNITE-20163
+            return new BaseTypeSpec(null, nullable);
+        } else if (SqlTypeUtil.isNull(type)) {
+            return RowSchemaTypes.NULL;
+        } else if (simpleType) {
+            NativeType nativeType = IgniteTypeFactory.relDataTypeToNative(type);
+            return RowSchemaTypes.nullableNativeType(nativeType, nullable);
+        } else if (type instanceof IntervalSqlType) {
+            IntervalSqlType intervalType = (IntervalSqlType) type;
+            boolean yearMonth = intervalType.getIntervalQualifier().isYearMonth();
+
+            if (yearMonth) {
+                // YEAR MONTH interval is stored as number of days in ints.
+                return RowSchemaTypes.nullableNativeType(NativeTypes.INT32, nullable);
+            } else {
+                // DAY interval is stored as time in nanos as long.
+                return RowSchemaTypes.nullableNativeType(NativeTypes.INT64, nullable);
+            }
+        } else if (SqlTypeUtil.isRow(type)) {
+            List<TypeSpec> fields = new ArrayList<>();
+
+            for (RelDataTypeField field : type.getFieldList()) {
+                TypeSpec fieldTypeSpec = convertToTypeSpec(field.getType());
+                fields.add(fieldTypeSpec);
+            }
+
+            return new RowType(fields, type.isNullable());
+
+        } else if (SqlTypeUtil.isMap(type) || SqlTypeUtil.isMultiset(type))  {
+            // TODO https://issues.apache.org/jira/browse/IGNITE-20162
+            throw new IllegalArgumentException("Collection types is not supported: " + type);
+        } else if (SqlTypeUtil.isArray(type)) {
+            // FIXME Aggregation type. Should be the same a branch with collection types.
```

Comment:
This can turned into TODO linked to issue  https://github.com/apache/ignite-3/pull/2312